File: src/main/java/shopping/token/application/TokenService.java
package shopping.token.application;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import shopping.member.application.MemberService;
import shopping.member.domain.Member;
import shopping.token.domain.Token;
import shopping.token.domain.TokenProvider;

@Service
public class TokenService {
    private final MemberService memberService;
    private final PasswordEncoder passwordEncoder;
    private final TokenProvider tokenProvider;

    public TokenService(MemberService memberService, PasswordEncoder passwordEncoder, TokenProvider tokenProvider) {
        this.memberService = memberService;
        this.passwordEncoder = passwordEncoder;
        this.tokenProvider = tokenProvider;
    }

    public Token generate(LoginRequest loginRequest) {
        Member member = memberService.getByEmail(loginRequest.email());
        validatePassword(loginRequest, member);
        return tokenProvider.generate(member.getEmail());
    }

    private void validatePassword(LoginRequest loginRequest, Member member) {
        if (isNotMatchedPassword(loginRequest, member)) {
            throw new NotMatchedPasswordException();
        }
    }

    private boolean isNotMatchedPassword(LoginRequest loginRequest, Member member) {
        return !passwordEncoder.matches(loginRequest.password(), member.getPassword());
    }

    public String extractEmail(String authorizationHeader) {
        return tokenProvider.extractEmail(authorizationHeader);
    }
}

File: src/main/kotlin/shopping/application/member/TokenService.kt
package shopping.application.member

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.security.Key
import java.util.Date

@Service
class TokenService(
    @Value("\${jwt.secret-key}") secretKeyString: String,
) {
    private val secretKey: Key = Keys.hmacShaKeyFor(secretKeyString.toByteArray())
    private val expirationMillis = 1000 * 60 * 60 // 1시간

    fun generateToken(
        memberId: Long,
        email: String,
    ): String {
        val now = Date()
        val expiry = Date(now.time + expirationMillis)

        return Jwts
            .builder()
            .setSubject(memberId.toString())
            .claim("email", email)
            .setIssuedAt(now)
            .setExpiration(expiry)
            .signWith(secretKey, SignatureAlgorithm.HS256)
            .compact()
    }

    fun validateToken(token: String): Boolean =
        try {
            Jwts
                .parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token)

            true
        } catch (e: Exception) {
            false
        }

    fun getMemberId(token: String): Long {
        val claims = getClaims(token)

        return claims.subject.toLong()
    }

    fun getEmail(token: String): String {
        val claims = getClaims(token)

        return claims["email"] as String
    }

    private fun getClaims(token: String): Claims =
        Jwts
            .parserBuilder()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body
}
